src/java.base/share/classes/com/sun/crypto/provider/XWing.java

```diff
@@ -0,0 +1,387 @@
+                return (PrivateKey) keyFactory.translateKey(key);
+            } catch (NoSuchAlgorithmException e) {
+                throw new AssertionError("SunJCE known to support ML-KEM", e);
+            } catch (InvalidKeyException e) {
+                throw new IllegalStateException("Implementation-internal key invalid", e);
+            }
+        }
+
+        @Override
+        public byte[] pkX() {
+            if (!(x.getPublic() instanceof XECPublicKey p)) {
+                throw new IllegalStateException("Unexpected type of X25519 public key: " + x.getPublic().getClass());
+            }
+
+            BigInteger u = p.getU();
+            byte[] uBytes = u.toByteArray();
+            ArrayUtil.reverse(uBytes);
+
+            return Arrays.copyOf(uBytes, 32); // Ensure it is 32 bytes long
+        }
+
+        /// returns the X25519 private key
+        /// @return the contained {@link PrivateKey} instance
+        public PrivateKey getX25519PrivateKey() {
+            return x.getPrivate();
+        }
+
+        @Override
+        public void destroy() {
+            Arrays.fill(m.decapsulationKey().keyBytes(), (byte) 0);
+            destroyQuietly(x.getPrivate());
+        }
+    }
+
+    // https://www.ietf.org/archive/id/draft-connolly-cfrg-xwing-kem-07.html#section-5.2
+    static XWingKeyPair expandDecapsulationKey(byte[] sk) {
+        assert sk.length == 32;
+
+        var expanded = shake256(sk, 96);
+        byte[] d = Arrays.copyOfRange(expanded, 0, 32);
+        byte[] z = Arrays.copyOfRange(expanded, 32, 64);
+        byte[] skX = Arrays.copyOfRange(expanded, 64, 96);
+        Arrays.fill(expanded, (byte) 0);
+
+        ML_KEM.ML_KEM_KeyPair m;
+        try {
+            m = new ML_KEM("ML-KEM-768").generateKemKeyPair(d, z);
+        } finally {
+            Arrays.fill(d, (byte) 0);
+            Arrays.fill(z, (byte) 0);
+        }
+
+        KeyPair x;
+        try {
+            x = deriveX25519KeyPair(skX);
+        } finally {
+            Arrays.fill(skX, (byte) 0);
+        }
+        return new XWingKeyPair(m, x);
+    }
+
+    // https://datatracker.ietf.org/doc/html/draft-connolly-cfrg-xwing-kem-07#section-5.3
+    private static byte[] combiner(byte[] ssM, byte[] ssX, byte[] ctX, byte[] pkX) {
+        var input = new byte[ssM.length + ssX.length + ctX.length + pkX.length + X_WING_LABEL.length];
+        System.arraycopy(ssM, 0, input, 0, ssM.length);
+        System.arraycopy(ssX, 0, input, ssM.length, ssX.length);
+        System.arraycopy(ctX, 0, input, ssM.length + ssX.length, ctX.length);
+        System.arraycopy(pkX, 0, input, ssM.length + ssX.length + ctX.length, pkX.length);
+        System.arraycopy(X_WING_LABEL, 0, input, ssM.length + ssX.length + ctX.length + pkX.length, X_WING_LABEL.length);
+        try {
+            MessageDigest sha3 = MessageDigest.getInstance("SHA3-256");
```

Comment:
@wangweij Shall I hardcode the SUN provider or leave it open for the implementation to find a suitable provider?